url9ecf1e8297_nhamidi_ninja_tgz-pJ8-net_fec_openrq_test_TestRunnerJ8/src/net/fec/openrq/core/util/rq/Utilities.java

```diff
@@ -0,0 +1,614 @@
+		}
+		*/
+
+		byte[] xor = new byte[length];
+		
+		for(int i=0; i<length; i++){
+			xor[i] = (byte) (s1[pos1+i] ^ s2[pos2+i]);
+		}
+		
+		return xor;
+	}
+
+	public static long ceilPrime(long p){
+
+		if(p == 1)
+			p++;
+		
+		while(!isPrime(p)) 
+			p++;
+
+		return p;
+	}
+
+	public static boolean isPrime(long n){
+		
+		//check if n is a multiple of 2
+		if (n % 2 == 0) return false;
+		
+		//if not, then just check the odds
+		for(long i = 3; i * i <= n; i += 2)
+			if(n % i == 0)
+				return false;
+		
+		return true;
+	}
+	
+	public static boolean checkIdentity(byte[][] A, int L) {
+		
+		for(int row = 0; row < L; row++)
+			for(int col = 0; col < L; col++)
+				if(row != col && A[row][col] != 0)
+					return false;
+				else if(row == col && A[row][col] != 1)
+					return false;
+		
+		return true;
+	}
+	
+	// works only for matrices in row echelon form
+		public static boolean validateRank(byte[][] matrix, int first_row, int first_col, int last_row, int last_col, int rank_lower_limit){
+			
+			int nonZeroRows = 0;
+			
+			for(int row = first_row; row < last_row; row++){
+				
+				for(int col = first_col; col < last_col; col++){
+
+					if(matrix[row][col] == 0)
+						continue;
+					else{
+						nonZeroRows++;
+						break;
+					}
+				}
+			}
+			
+			if(nonZeroRows < rank_lower_limit)
+				return false;
+			else
+				return true;
+		}
+
+		public static void rowEchelonForm(byte[][] A) throws SingularMatrixException{
+			
+			int ROWS = A.length;
+			int COLS = A[0].length;
+			
+			for(int row=0; row<COLS; row++){
+				
+				int max = row;
+				
+				// find pivot row and swap
+				for (int i = row + 1; i < ROWS; i++)
+					if (OctectOps.UNSIGN(A[i][row]) > OctectOps.UNSIGN(A[max][row]))
+						max = i;
+
+				// this destroys the original matrixes... dont really need a fix, but should be kept in mind
+				byte[] temp = A[row];
+				A[row] = A[max];
+				A[max] = temp;
+				
+				// singular
+	            if (A[row][row] == 0)
+					throw new SingularMatrixException();
+	 
+	            // pivot within A
+	            for(int i=row+1; i<ROWS; i++) {
+	            	
+	            	byte alpha = OctectOps.division(A[i][row], A[row][row]);
+	            	
+	            	for(int j=row; j<COLS; j++) {
+	            	
+	            		byte aux = OctectOps.product(alpha, A[row][j]);
+	            		
+	            		A[i][j] = OctectOps.subtraction(A[i][j], aux);
+	            	}
+	            }
+			}
+		}
+		
+		public boolean symbolIsZero(byte[] symbol){
+
+			for(byte b : symbol)
+				if(b != 0)
+					return false;
+
+			return true;
+		}
+		
+		/**
+		 * Solves a 'A.x = b' system using regular Gauss elimination.
+		 * @param A
+		 * @param b
+		 * @return Array of symbols 'x'
+		 * @throws SingularMatrixException
+		 */
+		public static byte[] gaussElimination(byte[][] A, byte[][] b) throws SingularMatrixException
+		{
+			if (A.length != A[0].length || b.length != A.length)
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem:**
 This line of code lacks validation when processing input data through the following parameter: &#39;b&#39; (index: 1 | type: byte[][). The parameter is exposed to external callers, because its enclosing class and method are publicly accessible. This means that upstream validation, if it exists, can be bypassed. Other validated parameters: &#39;A&#39;. The same parameter type is validated here for example: url9ecf1e8297_nhamidi_ninja_tgz-pJ8-net_fec_openrq_test_TestRunnerJ8/src/RQLibrary/Encoder.java:2059. Malicious, malformed, or unbounded inputs can cause unexpected runtime behavior or crashes, and can slow performance.

**Fix:**
 Add checks to ensure the validity of the parameter&#39;s value, such as testing it for nullness (for example, using the @NonNull annotation as [described in the Lombok library](https://projectlombok.org/features/NonNull)), emptiness, or equality. Or to prevent direct calls to it, [reduce the method&#39;s visibility](https://docs.oracle.com/javase/tutorial/java/javaOO/accesscontrol.html).

 Learn more about potential threats and guidance from the [Common Weakness Enumeration website](https://cwe.mitre.org/data/definitions/693.html) and the [OWASP Cheat Sheet series](https://cheatsheetseries.owasp.org/cheatsheets/Input_Validation_Cheat_Sheet.html).